lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClassCheck/SuperclassCycleCheck.hs

```diff
@@ -0,0 +1,98 @@
+module LambdaBuffers.Compiler.TypeClassCheck.SuperclassCycleCheck (runCheck) where
+
+import Control.Lens ((&), (.~), (^.))
+import Control.Monad (when)
+import Control.Monad.Error.Class (MonadError (throwError))
+import Control.Monad.Except (runExcept)
+import Control.Monad.Reader (ReaderT (runReaderT))
+import Control.Monad.Reader.Class (MonadReader, asks, local)
+import Data.Foldable (Foldable (foldl', toList), for_)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.ProtoLens (Message (defMessage))
+import Data.Set (Set)
+import Data.Set qualified as Set
+import LambdaBuffers.Compiler.ProtoCompat qualified as PC
+import Proto.Compiler qualified as P
+import Proto.Compiler_Fields (unboundClassRefErr)
+import Proto.Compiler_Fields qualified as P
+
+data CheckRead = MkCheckRead
+  { currentModuleName :: PC.ModuleName
+  , currentClassName :: PC.ClassName
+  , classDefs :: Map (PC.InfoLess PC.ModuleName, PC.InfoLess PC.ClassName) PC.ClassDef
+  , trace :: Set (PC.InfoLess PC.ClassName)
+  , reportingTrace :: [PC.TyClassRef]
+  }
+
+type MonadCheck m = (MonadReader CheckRead m, MonadError P.TyClassCheckError m)
+
+runCheck :: PC.CompilerInput -> Either [P.TyClassCheckError] ()
+runCheck ci =
+  let -- Index the class definitions by a qualified name (ModuleName, ClassName)
+      classDefs =
+        Map.fromList
+          [ ((mn, cn), cd)
+          | (mn, m) <- Map.toList $ ci ^. #modules
+          , (cn, cd) <- Map.toList $ m ^. #classDefs
+          ]
+      -- Process each type class definitions under the appropriate reader context and collect errors
+      allErrors =
+        foldl'
+          ( \errs (mn, cd) ->
+              let errM = runReaderT (checkClassDef cd) (MkCheckRead mn (cd ^. #className) classDefs Set.empty [])
+               in case runExcept errM of
+                    Left err -> err : errs
+                    Right _ -> errs
+          )
+          []
+          [(m ^. #moduleName, cd) | m <- toList $ ci ^. #modules, cd <- toList $ m ^. #classDefs]
+   in if null allErrors then Right () else Left allErrors
+
+checkClassDef :: MonadCheck m => PC.ClassDef -> m ()
+checkClassDef (PC.ClassDef cn _ sups _ _) = do
+  checkCycle cn
```

Comment:
Make it clear you don't use the result - use `void` or `_ <-`.

Reply:
The result is () so that makes it clear imo